libqrexec/process_io.c

```diff
@@ -40,45 +40,40 @@ static _Noreturn void handle_vchan_error(const char *op)
 /*
  * Closing the file descriptors:
  *
- * - Use shutdown(), except if this is a FD inherited from parent (detected by
- *   FD number)
- * - Restore blocking status, unless this is a duplicated stdio socket for
- *   stdout/stderr (in which case we can't restore it just for one FD, but we
- *   don't care, because we created it)
+ * - If this is a single socket used for both stdin and stdout, call shutdown()
+ *   to indicate that no more data will be sent or received.
+ * - Otherwise, restore the blocking status of the FD.
  */
 
-static void close_stdin(int fd, bool restore_block) {
+static void close_stdin(int fd, bool restore_block, bool close_fd) {
     if (fd == -1)
         return;
 
     if (restore_block)
         set_block(fd);
-
-    if (fd == 1) {
-        close(fd);
-    } else if (shutdown(fd, SHUT_WR) == -1) {
-        if (errno == ENOTSOCK)
-            close(fd);
-        else
+    else {
+        if (shutdown(fd, SHUT_WR) == -1)
             PERROR("shutdown close_stdin");
     }
+
+    if (close_fd)
+        close(fd);
```

Comment:
Now, with restore_block=true and close_fd=false this function won't close (using any method) the FD at all. I haven't checked if such combination is possible, but this behavior feels wrong: if it can happen, then it needs to close the VM using some method. If not, documentation above wants an update (and maybe also include an assert about this?). And then the parameter name is confusing, as `restore_block` now also controls using `shutdown` or not.

Reply:
And also, previous version handled ENOTSOCK. Are you absolutely sure it can't happen? `process_io` may get the FD from the user process, so I don't think you can assume it's always a socket (even if this path would be reachable only after service sending SIGUSR1).

Reply:
> Now, with restore_block=true and close_fd=false this function won't close (using any method) the FD at all. I haven't checked if such combination is possible, but this behavior feels wrong: if it can happen, then it needs to close the VM using some method. If not, documentation above wants an update (and maybe also include an assert about this?).

This _will_ happen for socket-based services, but the behavior is correct: the call to `close_stdin()` is effectively a no-op, and the call to `close_stdout()` closes the socket.

<https://github.com/QubesOS/qubes-core-qrexec/blob/d1f525204c5004c23fc7557433df33b3edb203cd/libqrexec/process_io.c#L307-L308>

However, I completely rewrote this in the most recent force-push.

> And then the parameter name is confusing, as `restore_block` now also controls using `shutdown` or not.

Fixed by getting rid of the parameter.

> And also, previous version handled ENOTSOCK. Are you absolutely sure it can't happen? `process_io` may get the FD from the user process, so I don't think you can assume it's always a socket (even if this path would be reachable only after service sending SIGUSR1).

Fixed.